design/230728_tinkerbell_workflow_execution.md

```diff
@@ -0,0 +1,491 @@
+
+Tink Worker communicates with Tink Server over gRPC. There are a handful of obsolete RPCs that are unimplemented in Tink Server since the transition to the Kubernetes backend (see Appendix A for existing protobuf).
+
+## Proposal
+
+### Summary
+
+The components used in Workflow execution will remain similar to the existing solution. Tink Agent (renamed from Tink Worker) will run on the device to be provisioned and is responsible for executing Workflows. Tink Server will instruct Tink Agent to run and cancel Workflows, and will update Workflow status based on Workflow execution Events published by Tink Agent. Tink Controller will prepare Workflows for execution and monitor them for erroneous states taking any necessary remedial action. The architecture draws similarities with the [Scheduler-Agent-Supervisor](https://learn.microsoft.com/en-us/azure/architecture/patterns/scheduler-agent-supervisor) pattern.
+
+### Workflow States
+
+<p align="center">
+  <img alt="[State Transitions]" src="./images/230728_tinkerbell_workflow_execution/state.png" />
+</p>
+
+#### Why do we need the Scheduled state?
+
+The Scheduled state is an intermediary state that indicates a Workflow has been dispatched to Tink Agent but hasn't started. Tink Controller will monitor Workflows in a Scheduled state and transition them to Failed if a timeout trips.
+
+Workflows may also be rejected by Tink Agent explicitly for cases such as a double dispatch. In the event a Workflow is rejected it will transition back to Pending. Workflow rejection is discussed later.
+
+#### Why do we need the Cancelling state?
+
+The Cancelling state separates Workflows we intend to cancel from those that have been cancelled. This allows Tink Server to issue a cancellation instruction to Tink Agent, and for Tink Agent to indicate successful cancellation. The explicitness of this sequence is useful because:
+* Explicit indication that a Running Workflow has stopped and thus Tink Agent is availbale for a subsequent Workflow is important to avoid unintended Workflow rejection (double dispatch).
+* It aids in avoiding a double dispatch. Canceled is an end state so the system is at liberty to dispatch a subsequent Workflow even if the previous Workflow is still in the process of cancelling.
+* It makes clear to operators the state of the Workflow and why Tink Agent isn't executing a subsequent Workflow.
+
+### Executing Workflows
+
+Tink Server will continue to expose a gRPC API. The protobuf will be simplified to remove superfluous APIs and data structures. Tink Agent will connect to Tink Server to create a command stream. When Tink Server identifies a runnable Workflow for a connected Tink Agent it will dispatch the Workflow on the command stream. Workflows are dispatched in the order they are created. As Tink Agent executes Workflow Actions it will publish events to Tink Server.
+
+<p align="center">
+  <img src="images/230728_tinkerbell_workflow_execution/executing-a-workflow.png" />
```

Comment:
We should also include newly added states like "scheduled", "cancelling", "cancelled", "succeeded" to the sequence diagram.